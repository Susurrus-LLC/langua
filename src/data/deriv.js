export const defData = {
  words: 5,
  type: 'any',
  lexemes: [
    {
      lexeme: 'nzi',
      definition: 'tree'
    },
    {
      lexeme: 'ilu',
      definition: 'flower'
    },
    {
      lexeme: 'ru',
      definition: 'person'
    },
    {
      lexeme: 'kwərə',
      definition: 'goat'
    },
    {
      lexeme: 'lɛtsi',
      definition: 'sword'
    }
  ],
  derivations: [
    {
      derivation: '-vo',
      gloss: 'person',
      definition:
        'Creates a noun referring to a person associated with the base.'
    },
    {
      derivation: '-lu',
      gloss: 'place',
      definition: 'Creates a noun denoting the location of the base.'
    },
    {
      derivation: 'ña-',
      gloss: 'abstract',
      definition: 'Creates an abstract noun.'
    },
    {
      derivation: 'k-sa',
      gloss: 'collective',
      definition: 'Creates a noun referring to a collection of bases.'
    }
  ],
  results: undefined
}
