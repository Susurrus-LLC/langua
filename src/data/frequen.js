export const defData = {
  corpus:
    'Nzɔ de’áxa Tsənaxa, ño lo sɛtsa’ɔk, ño lo de’áxa u Kasaga, ño lo mɛya ayru Tol o Əbe kwə, sa nzɔ xa mpaskale: Steyə rabɛ lu ñire lu mvunə poxe o lu pɛlɔ ah lo mɛwɛ axe, maxə ɔra latsi əbɔwə səyo.',
  consonants:
    'b/d/f/g/k/kw/l/m/mp/mv/n/nt/nts/nz/ñ/ŋk/ŋkw/p/r/s/t/ts/v/w/x,h/y/’',
  vowels: 'a,á/ɛ,ɛ́/e,é/ə,ə́/i,í/ɔ,ɔ́/o,ó/u,ú',
  distinguishCase: false,
  analyzed: false,
  results: undefined,
  hovered: undefined,
  filterSeg: 'all',
  filterSyll: 'all',
  filterWords: 'all'
}
