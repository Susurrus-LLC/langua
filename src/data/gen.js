export const defData = {
  subpatterns: [
    {
      selected: 'V',
      subpattern: 'a/e/i/o/u'
    },
    {
      selected: 'C',
      subpattern: 'p/t/k/b/d/g/s/m/n/l/r'
    },
    {
      selected: 'N',
      subpattern: 'm/n'
    }
  ],
  pattern: 'V/CV/CVN/CVCV/CVCVN',
  words: 100,
  newline: false,
  filterdupes: false
}

export const vars = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z'
]
