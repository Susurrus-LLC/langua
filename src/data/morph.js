export const defData = {
  categories: [
    'V=ieaou',
    'L=īēāōū',
    'C=ptcqbdgmnlrhs',
    'F=ie',
    'B=ou',
    'S=ptc',
    'Z=bdg'
  ],
  rewriteRules: ['lh=lj'],
  soundChanges: [
    '[sm]//_#',
    'i/j/_V',
    'L/V/_',
    'e//Vr_#',
    'v//V_V',
    'u/o/_#',
    'gn/nh/_',
    'S/Z/V_V',
    'c/i/F_t',
    'c/u/B_t',
    'p//V_t',
    'ii/i/_',
    'e//C_rV'
  ],
  lexicon: [
    'lector',
    'doctor',
    'focus',
    'jocus',
    'districtus',
    'cīvitatem',
    'adoptare',
    'opera',
    'secundus',
    'fīliam',
    'pōntem'
  ],
  outputFormat: 'oo',
  showDiff: true,
  showChanges: true,
  ruleReport: true,
  rewriteOutput: true,
  results: undefined
}
