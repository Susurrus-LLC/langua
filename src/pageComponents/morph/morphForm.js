import React from 'react'
import PropTypes from 'prop-types'

import Form from '../../components/form'

import sharedStyles from '../../components/form/sharedForm.module.sass'

const MorphForm = () => <Form name='morph-form' />

MorphForm.propTypes = {
  styles: PropTypes.string
}

export default MorphForm
