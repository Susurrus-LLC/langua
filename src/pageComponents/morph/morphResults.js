import React from 'react'
import PropTypes from 'prop-types'

import Results from '../../components/results'

import sharedStyles from '../../components/results/sharedResults.module.sass'

const MorphResults = () => <Results />

MorphResults.propTypes = {
  styles: PropTypes.string
}

export default MorphResults
