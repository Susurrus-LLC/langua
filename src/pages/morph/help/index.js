import React from 'react'
import { graphql, useStaticQuery } from 'gatsby'

import Layout from '../../../components/layout'
import SEO from '../../../components/seo'
import Notice from '../../../components/notice'
import Help from '../../../components/help'

const MorphHelpPage = () => {
  const data = useStaticQuery(graphql`
    query MorphHelpQuery {
      site {
        siteMetadata {
          title
          toolInfo {
            morph {
              title
              link
              description
            }
          }
        }
      }
    }
  `)

  const metadata = data.site.siteMetadata
  const toolInfo = metadata.toolInfo.morph

  return (
    <Layout>
      <SEO
        title={`${metadata.title}${toolInfo.title} Help`}
        canonical={`${toolInfo.link}/help`}
      />
      <Help toolInfo={toolInfo}>
        <Notice>This tool is still in development.</Notice>
        <h3 id='using'>
          Using {metadata.title}
          {toolInfo.title}
        </h3>
        <h3 id='acknowledgments'>Acknowledgments</h3>
        <p>
          Much thanks should be given to Mark Rosenfelder and{' '}
          <a
            href='http://www.zompist.com/sca2.html'
            target='_blank'
            rel='noopener noreferrer'
          >
            the Sound Change Applier 2
          </a>{' '}
          (SCA<sup>2</sup>). {metadata.title}
          {toolInfo.title} was mainly built as a modernized and updated version
          of SCA<sup>2</sup>.
        </p>
      </Help>
    </Layout>
  )
}

export default MorphHelpPage
