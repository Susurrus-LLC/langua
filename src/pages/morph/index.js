import React, { useState } from 'react'
import { graphql, useStaticQuery } from 'gatsby'

import Notice from '../../components/notice'
import Tool from '../../components/tool'

import MorphForm from '../../pageComponents/morph/morphForm'
import MorphResults from '../../pageComponents/morph/morphResults'
import morphService from '../../services/morphService'

import styles from './morph.module.sass'

export const PureMorph = ({ data }) => {
  const toolInfo = data?.site?.siteMetadata?.toolInfo?.morph

  // State

  const [categories, setCategories] = useState(
    morphService.getData().categories
  )
  const [rewriteRules, setRewriteRules] = useState(
    morphService.getData().rewriteRules
  )
  const [soundChanges, setSoundChanges] = useState(
    morphService.getData().soundChanges
  )
  const [lexicon, setLexicon] = useState(morphService.getData().lexicon)
  const [outputFormat, setOutputFormat] = useState(
    morphService.getData().outputFormat
  )
  const [showDiff, setShowDiff] = useState(morphService.getData().showDiff)
  const [showChanges, setShowChanges] = useState(
    morphService.getData().showChanges
  )
  const [ruleReport, setRuleReport] = useState(
    morphService.getData().ruleReport
  )
  const [rewriteOutput, setRewriteOutput] = useState(
    morphService.getData().rewriteOutput
  )
  const [results, setResults] = useState(morphService.getData().results)

  // Functions

  const onChange = e => {
    // When an input field is changed, update state
    const name = e.target.name
    let val
    if (e.target.type === 'textarea') {
      // Split the textarea inputs by newlines
      val = e.target.value.split('\n')
    } else {
      // Just get the value of the input
      val = e.target.value
    }

    switch (name) {
      case 'categories':
        setCategories(val)
        break
      case 'rewriteRules':
        setRewriteRules(val)
        break
      case 'soundChanges':
        setSoundChanges(val)
        break
      case 'lexicon':
        setLexicon(val)
        break
      case 'outputFormat':
        setOutputFormat(val)
        break
      default:
    }
  }

  const onCheck = e => {
    // When a checkbox option is changed, update state
    const val = e.target.value
    const checked = e.target.checked
    switch (val) {
      case 'showChanges':
        setShowChanges(checked)
        break
      case 'showDiff':
        setShowDiff(checked)
        break
      case 'ruleReport':
        setRuleReport(checked)
        break
      case 'rewriteOutput':
        setRewriteOutput(checked)
        break
      default:
    }
  }

  const onMorph = e => {
    e.preventDefault()
    const state = {
      categories,
      rewriteRules,
      soundChanges,
      lexicon,
      outputFormat,
      showDiff,
      showChanges,
      ruleReport,
      rewriteOutput,
      results
    }
    // Generate the output
    setResults(morphService.morph(state))
    // Save the current state to storage
    morphService.setStorage(state)
  }

  const onSave = e => {
    e.preventDefault()
    const state = {
      categories,
      rewriteRules,
      soundChanges,
      lexicon,
      outputFormat,
      showDiff,
      showChanges,
      ruleReport,
      rewriteOutput,
      results
    }
    // Save the current state to storage and generate a file
    morphService.save(state)
  }

  const onOpen = e => {
    e.preventDefault()
    const file = e.target.files[0]
    const updateState = response => {
      // Only change state if the file was successfully opened
      if (response) {
        setCategories(response.categories)
        setRewriteRules(response.rewriteRules)
        setSoundChanges(response.soundChanges)
        setLexicon(response.lexicon)
        setOutputFormat(response.outputFormat)
        setShowDiff(response.showDiff)
        setShowChanges(response.showChanges)
        setRuleReport(response.ruleReport)
        setRewriteOutput(response.rewriteOutput)
        setResults(response.results)
      }
    }

    morphService.open(file, updateState)
  }

  return (
    <Tool toolInfo={toolInfo}>
      <Notice>This tool is still in development.</Notice>
      <MorphForm
        styles={styles}
        data={{
          categories,
          rewriteRules,
          soundChanges,
          lexicon,
          outputFormat,
          showDiff,
          showChanges,
          ruleReport,
          rewriteOutput
        }}
        change={onChange}
        morph={onMorph}
        check={onCheck}
        save={onSave}
        open={onOpen}
      />
      <MorphResults
        styles={styles}
        outputFormat={outputFormat}
        results={results}
        showDiff={showDiff}
        showChanges={showChanges}
      />
    </Tool>
  )
}

const Morph = props => {
  const data = useStaticQuery(graphql`
    query MorphToolQuery {
      site {
        siteMetadata {
          toolInfo {
            morph {
              title
              link
              description
            }
          }
        }
      }
    }
  `)

  return <PureMorph {...props} data={data} />
}

export default Morph
