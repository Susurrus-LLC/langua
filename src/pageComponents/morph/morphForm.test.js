import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import MorphForm from './morphForm'
import { defData } from '../../data/morph'

describe('MorphForm', () => {
  const styles = {
    section: 'section',
    categories: 'categories',
    inputSection: 'inputSection',
    container: 'container',
    categoriesInput: 'categoriesInput',
    dataInput: 'dataInput',
    textAreaSpacer: 'textAreaSpacer',
    rewriteRules: 'rewriteRules',
    rewriteRulesInput: 'rewriteRulesInput',
    soundChanges: 'soundChanges',
    sectionTitle: 'sectionTitle',
    soundChangesInput: 'soundChangesInput',
    lexicon: 'lexicon',
    lexiconInput: 'lexiconInput'
  }

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <MorphForm
        styles={styles}
        data={defData}
        change={jest.fn()}
        morph={jest.fn()}
        check={jest.fn()}
        save={jest.fn()}
        open={jest.fn()}
      />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders accurately', () => {
    let tree = renderer
      .create(
        <MorphForm
          styles={styles}
          data={defData}
          change={jest.fn()}
          morph={jest.fn()}
          check={jest.fn()}
          save={jest.fn()}
          open={jest.fn()}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
