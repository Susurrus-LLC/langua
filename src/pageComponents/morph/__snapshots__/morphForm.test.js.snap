// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MorphForm renders accurately 1`] = `
<form
  className="form"
  name="morph-form"
>
  <div
    className="section"
  >
    <div
      className="categories inputSection part"
    >
      <h5
        className="sectionTitle"
      >
        Categories
      </h5>
      <div
        className="container"
      >
        <textarea
          aria-label="Categories"
          className="categoriesInput dataInput"
          id="categories"
          name="categories"
          onChange={[MockFunction]}
          value="V=ieaou
L=īēāōū
C=ptcqbdgmnlrhs
F=ie
B=ou
S=ptc
Z=bdg"
        />
        <div
          className="textAreaSpacer"
        >
          V=ieaou
L=īēāōū
C=ptcqbdgmnlrhs
F=ie
B=ou
S=ptc
Z=bdg
           
        </div>
      </div>
    </div>
    <div
      className="rewriteRules inputSection part"
    >
      <h5
        className="sectionTitle"
      >
        Rewrite Rules
      </h5>
      <div
        className="container"
      >
        <textarea
          aria-label="Rewrite rules"
          className="rewriteRulesInput dataInput"
          id="rewriteRules"
          name="rewriteRules"
          onChange={[MockFunction]}
          value="lh=lj"
        />
        <div
          className="textAreaSpacer"
        >
          lh=lj
           
        </div>
      </div>
    </div>
    <div
      className="soundChanges inputSection part"
    >
      <h5
        className="sectionTitle"
      >
        Sound Changes
      </h5>
      <div
        className="container"
      >
        <textarea
          aria-label="Sound changes"
          className="soundChangesInput dataInput"
          id="soundChanges"
          name="soundChanges"
          onChange={[MockFunction]}
          value="[sm]//_#
i/j/_V
L/V/_
e//Vr_#
v//V_V
u/o/_#
gn/nh/_
S/Z/V_V
c/i/F_t
c/u/B_t
p//V_t
ii/i/_
e//C_rV"
        />
        <div
          className="textAreaSpacer"
        >
          [sm]//_#
i/j/_V
L/V/_
e//Vr_#
v//V_V
u/o/_#
gn/nh/_
S/Z/V_V
c/i/F_t
c/u/B_t
p//V_t
ii/i/_
e//C_rV
           
        </div>
      </div>
    </div>
    <div
      className="lexicon inputSection part"
    >
      <h5
        className="sectionTitle"
      >
        Lexicon
      </h5>
      <div
        className="container"
      >
        <textarea
          aria-label="Lexicon"
          className="lexiconInput dataInput"
          id="lexicon"
          name="lexicon"
          onChange={[MockFunction]}
          value="lector
doctor
focus
jocus
districtus
cīvitatem
adoptare
opera
secundus
fīliam
pōntem"
        />
        <div
          className="textAreaSpacer"
        >
          lector
doctor
focus
jocus
districtus
cīvitatem
adoptare
opera
secundus
fīliam
pōntem
           
        </div>
      </div>
    </div>
  </div>
  <div
    className="control part"
  >
    <div
      className="controlSide controlLeft"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn neutral"
          onClick={[MockFunction]}
          role="button"
          type="submit"
        >
          Change
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <label
          htmlFor="outputFormat"
        >
          output format:
        </label>
        <select
          id="outputFormat"
          name="outputFormat"
          onChange={[MockFunction]}
          value="oo"
        >
          <option
            value="oo"
          >
            output
          </option>
          <option
            value="io"
          >
            input → output
          </option>
          <option
            value="oi"
          >
            output ← input
          </option>
        </select>
      </div>
      <div
        className="controlPiece"
      >
        <div>
          <label>
            <input
              checked={true}
              id="showChanges"
              name="options"
              onChange={[MockFunction]}
              type="checkbox"
              value="showChanges"
            />
             
            italicize changes
          </label>
          <br />
          <label>
            <input
              checked={true}
              id="showDiff"
              name="options"
              onChange={[MockFunction]}
              type="checkbox"
              value="showDiff"
            />
             
            bold differences
          </label>
        </div>
      </div>
      <div
        className="controlPiece"
      >
        <div>
          <label>
            <input
              checked={true}
              id="ruleReport"
              name="options"
              onChange={[MockFunction]}
              type="checkbox"
              value="ruleReport"
            />
             
            report applied rules
          </label>
          <br />
          <label>
            <input
              checked={true}
              id="rewriteOutput"
              name="options"
              onChange={[MockFunction]}
              type="checkbox"
              value="rewriteOutput"
            />
             
            rewrite on output
          </label>
        </div>
      </div>
    </div>
    <div
      className="controlSide controlRight"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn success"
          id="save"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Save
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <button
          className="btn"
          id="open"
          onClick={[Function]}
          role="button"
          type="button"
        >
          Open
        </button>
        <input
          className="hidden"
          id="file"
          name="file"
          onChange={[MockFunction]}
          type="file"
        />
      </div>
    </div>
  </div>
</form>
`;
